employee-rostering-frontend/src/ui/header/Toolbar.tsx

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Button,
+  ButtonVariant,
+  Dropdown,
+  DropdownItem,
+  DropdownToggle,
+  Toolbar,
+  ToolbarGroup,
+  ToolbarItem,
+} from '@patternfly/react-core';
+import { BellIcon, CogIcon } from '@patternfly/react-icons';
+import * as React from 'react';
+import { connect } from 'react-redux';
+import { tenantOperations } from 'store/tenant';
+import { AppState } from 'store/types';
+import Tenant from 'domain/Tenant';
+import TenantRestServiceClient from 'services/TenantRestServiceClient';
+
+interface StateProps {
+  currentTenantId: number;
+  tenantList: Tenant[];
+}
+
+const mapStateToProps = ({ tenantData }: AppState): StateProps => ({
+  currentTenantId: tenantData.currentTenantId,
+  tenantList: tenantData.tenantList
+});
+
+interface ToolbarState {
+  isTenantSelectOpen: boolean;
+}
+
+export interface DispatchProps {
+  refreshTenantList: typeof tenantOperations.refreshTenantList;
+  changeTenant: typeof tenantOperations.changeTenant;
+}
+
+const mapDispatchToProps: DispatchProps = {
+  refreshTenantList: tenantOperations.refreshTenantList,
+  changeTenant: tenantOperations.changeTenant
+};
+
+export type Props = StateProps & DispatchProps;
+
+class ToolbarComponent extends React.Component<Props, ToolbarState> {
+  tenantRestService: TenantRestServiceClient = new TenantRestServiceClient();
+  toolbarState: ToolbarState;
```

Comment:
This really shouldn't be there. Since you've defined the shape of the components state in `extends React.Component<Props, State>`, the state is already held by parent's `state` property. So you can access it for read as `this.state.isTenantSelectOpen` (`this.state` is of type `ToolbarState`) and `this.setState({isTenantSelectOpen: true})` for write.

Reply:
Reading is tricky; `state` is not updated until the component is re-rendered. This turns out to be an issue for the first render; I get "Cannot read property 'isTenantSelectOpen' of undefined" error when I set state in the constructor using `this.state` and update state using `this.setState` everywhere else.

Reply:
`this.toolbarState` was a workaround for reading the state on the first render; another solution is having a `const initalState = {isTenantSelectOpen : false}`, using `setState()` to update state, and in `render()`, have a constant `const state : ToolbarState = this.state || initalState`.

Reply:
Yeah, sure, it's necessary to initialize `this.state` in the constructor. You don't even have to have `const initialState`. It can be simplified to
```ts
constructor(props: Props) {
  super(props);
  this.state = { isTenantSelectOpen: false };
}
```